File: sdk/typescript/src/services/checkout_links.ts
import type { ClientBase } from "../base";
import type { CheckoutLinkCreate, CheckoutLinkUpdate } from "../models/inputs";
import type { CheckoutLinkSortProperty } from "../models/literals";
import type { CheckoutLink, ListResourceCheckoutLink } from "../models/outputs";

import { HTTPValidationError, ResourceNotFound } from "../errors";

export const listCheckoutLinks = (client: ClientBase) => {
  /**
   * List checkout links.
   *
   * **Scopes**: `checkout_links:read` `checkout_links:write`
   *
   * @param query - Query parameters
   * @returns {ListResourceCheckoutLink}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    organization_id?: string | string[] | null;
    product_id?: string | string[] | null;
    page?: number;
    limit?: number;
    sorting?: CheckoutLinkSortProperty[] | null;
  }): Promise<ListResourceCheckoutLink> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
      product_id: query?.product_id,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["created_at"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/checkout-links/",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourceCheckoutLink>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
/**
 * List checkout links.
 *
 * **Scopes**: `checkout_links:read` `checkout_links:write`
 *
 * @param query - Query parameters
 * @returns {AsyncGenerator<CheckoutLink>} A generator that yields items of type CheckoutLink.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {HTTPValidationError} Validation Error
 */
export const iterlistCheckoutLinks = (client: ClientBase) => {
  return async function* (query?: {
    organization_id?: string | string[] | null;
    product_id?: string | string[] | null;
    page?: number;
    limit?: number;
    sorting?: CheckoutLinkSortProperty[] | null;
  }): AsyncGenerator<CheckoutLink> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listCheckoutLinks(client)({ ...query, page, limit });
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const createCheckoutLinks = (client: ClientBase) => {
  /**
   * Create a checkout link.
   *
   * **Scopes**: `checkout_links:write`
   *
   * @param body - Request body
   * @returns {CheckoutLink}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body: CheckoutLinkCreate): Promise<CheckoutLink> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/checkout-links/",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutLink>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const getCheckoutLinks = (client: ClientBase) => {
  /**
   * Get a checkout link by ID.
   *
   * **Scopes**: `checkout_links:read` `checkout_links:write`
   *
   * @param id - The checkout link ID.
   * @returns {CheckoutLink}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Checkout link not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<CheckoutLink> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/checkout-links/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutLink>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const deleteCheckoutLinks = (client: ClientBase) => {
  /**
   * Delete a checkout link.
   *
   * **Scopes**: `checkout_links:write`
   *
   * @param id - The checkout link ID.
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Checkout link not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<void> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "DELETE",
      "/v1/checkout-links/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<void>(response, "none", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const updateCheckoutLinks = (client: ClientBase) => {
  /**
   * Update a checkout link.
   *
   * **Scopes**: `checkout_links:write`
   *
   * @param id - The checkout link ID.
   * @param body - Request body
   * @returns {CheckoutLink}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Checkout link not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string, body: CheckoutLinkUpdate): Promise<CheckoutLink> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/checkout-links/{id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutLink>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};

export function createCheckoutLinksService(client: ClientBase) {
  return {
    list: listCheckoutLinks(client),
    create: createCheckoutLinks(client),
    get: getCheckoutLinks(client),
    delete: deleteCheckoutLinks(client),
    update: updateCheckoutLinks(client),
    iterlist: iterlistCheckoutLinks(client),
  };
}

export type CheckoutLinks = ReturnType<typeof createCheckoutLinksService>;

File: sdk/python/polar/v2026_04/services/checkout_links.py
from __future__ import annotations

import builtins
import typing

from polar.base import (
    AsyncServiceBase,
    RequestTimeout,
    SyncServiceBase,
    parse_response_json,
    parse_response_none,
)
from polar.v2026_04.errors import (
    HTTPValidationError,
    ResourceNotFound,
)
from polar.v2026_04.inputs import (
    CheckoutLinkCreateProduct,
    CheckoutLinkCreateProductPrice,
    CheckoutLinkCreateProducts,
    CheckoutLinkUpdate,
)
from polar.v2026_04.literals import (
    CheckoutLinkSortProperty,
)
from polar.v2026_04.outputs import (
    CheckoutLink,
    ListResourceCheckoutLink,
)


class CheckoutLinksSync(SyncServiceBase):
    def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutLinkSortProperty] | None = ["created_at"],
        request_timeout: RequestTimeout | None = None,
    ) -> ListResourceCheckoutLink:
        """
        List checkout links.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkout-links/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "product_id": product_id,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
            request_timeout=request_timeout,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, ListResourceCheckoutLink, method_errors)

    def iter_list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutLinkSortProperty] | None = ["created_at"],
        request_timeout: RequestTimeout | None = None,
    ) -> typing.Generator[CheckoutLink, None, None]:
        """
        List checkout links.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            request_timeout: Timeout override for each request, in seconds or as an httpx.Timeout instance.


        Returns:
            A generator that yields items of type CheckoutLink.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = self.list(
                organization_id=organization_id,
                product_id=product_id,
                page=page,
                limit=limit,
                sorting=sorting,
                request_timeout=request_timeout,
            )
            yield from response.items
            if page >= response.pagination.max_page:
                break
            page += 1

    @typing.overload
    def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProductPrice],
    ) -> CheckoutLink: ...

    @typing.overload
    def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProduct],
    ) -> CheckoutLink: ...

    @typing.overload
    def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProducts],
    ) -> CheckoutLink: ...

    def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Any,
    ) -> CheckoutLink:
        """
        Create a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.

            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkout-links/",
            path_params={},
            query_params={},
            request_timeout=request_timeout,
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)

    def get(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
    ) -> CheckoutLink:
        """
        Get a checkout link by ID.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)

    def delete(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
    ) -> None:
        """
        Delete a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    def update(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkUpdate],
    ) -> CheckoutLink:
        """
        Update a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.

            **kwargs: Request body parameters

        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)


class CheckoutLinksAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutLinkSortProperty] | None = ["created_at"],
        request_timeout: RequestTimeout | None = None,
    ) -> ListResourceCheckoutLink:
        """
        List checkout links.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkout-links/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "product_id": product_id,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
            request_timeout=request_timeout,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, ListResourceCheckoutLink, method_errors)

    async def iter_list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutLinkSortProperty] | None = ["created_at"],
        request_timeout: RequestTimeout | None = None,
    ) -> typing.AsyncGenerator[CheckoutLink, None]:
        """
        List checkout links.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            request_timeout: Timeout override for each request, in seconds or as an httpx.Timeout instance.


        Returns:
            An async generator that yields items of type CheckoutLink.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = await self.list(
                organization_id=organization_id,
                product_id=product_id,
                page=page,
                limit=limit,
                sorting=sorting,
                request_timeout=request_timeout,
            )
            for item in response.items:
                yield item
            if page >= response.pagination.max_page:
                break
            page += 1

    @typing.overload
    async def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProductPrice],
    ) -> CheckoutLink: ...

    @typing.overload
    async def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProduct],
    ) -> CheckoutLink: ...

    @typing.overload
    async def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkCreateProducts],
    ) -> CheckoutLink: ...

    async def create(
        self,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Any,
    ) -> CheckoutLink:
        """
        Create a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.

            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkout-links/",
            path_params={},
            query_params={},
            request_timeout=request_timeout,
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)

    async def get(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
    ) -> CheckoutLink:
        """
        Get a checkout link by ID.

        **Scopes**: `checkout_links:read` `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)

    async def delete(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
    ) -> None:
        """
        Delete a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.


        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    async def update(
        self,
        id: str,
        *,
        request_timeout: RequestTimeout | None = None,
        **kwargs: typing.Unpack[CheckoutLinkUpdate],
    ) -> CheckoutLink:
        """
        Update a checkout link.

        **Scopes**: `checkout_links:write`

        Args:
            id: The checkout link ID.
            request_timeout: Timeout override for this request, in seconds or as an httpx.Timeout instance.

            **kwargs: Request body parameters

        Raises:
            ResourceNotFound: Checkout link not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkout-links/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            request_timeout=request_timeout,
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_json(response, CheckoutLink, method_errors)
